### PR TITLE
fix: node callbacks

### PR DESCRIPTION
**Pull request by lifeiscontent**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/api/fs.html#fs_fs_readfile_path_options_callback read at bottom where it says => null
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by lifeiscontent** [comment]
the only thing I did here was find/replace callbacks in node to make it so (err, data) => void doesn't always return an Error

**Review by SimonSchick — CHANGES_REQUESTED** [review]
Please revert the formatting changes, this diff is unreadable.

**Comment by lifeiscontent** [comment]
@SimonSchick updated

**Comment by Flarna** [comment]
@sandersn I think something is wrong with Azure build as it is green but should be red here.

File: types/node/index.d.ts
Comment (SimonSchick) [inline types/node/index.d.ts]: Iirc we cannot do this as it would break too many packages?

**Review by SimonSchick — CHANGES_REQUESTED** [review]
There are still too many formatting(quotestyle, spacing, newlines etc.) changes, it is extremely hard to distinguish functional changes from formatting changes.

Please revert all the formatting changes and create another PR for them.

Also it might be worth creating a `NullableErrno` or `NullableError` to avoid the constant type union creation.

File: types/node/index.d.ts
Comment (Flarna) [reply types/node/index.d.ts]: this breaks at least several hundreds packages and I don't think the `| null` changes require this.

**Comment by sandersn** [comment]
@Flarna I noticed that this morning on another PR. I'll see if I can figure out what is going on.

**Comment by sandersn** [comment]
@Flarna OK, Azure should be correctly reporting breaks again.

File: types/node/index.d.ts
Comment (lifeiscontent) [reply types/node/index.d.ts]: @SimonSchick sorry, I was only following what the linter was telling me to do. I'll force push over this with only changes to the null values, by implementing a Nullable

Comment (lifeiscontent) [reply types/node/index.d.ts]: @SimonSchick do you have a suggestion where I should implement the `NullableErrno`?

**Comment by lifeiscontent** [comment]
@SimonSchick I've updated the definitions with only the bare minimum changes.

**Review by SimonSchick — APPROVED** [review]

**Review by BrunoScheufler — APPROVED** [review]

**Comment by lifeiscontent** [comment]
@SimonSchick any updates?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by SimonSchick** [comment]
Changes look ok, but I think we should remove the changes from `v0`, `v4`, `v7` and `v9` as they are EOL.

**Comment by Flarna** [comment]
@lifeiscontent CI fails as your changes effect modules depending on node.
You either have to adapt them also or revert the corresponding changes in node otherwise this PR will be not merged as CI is red.

**Comment by lifeiscontent** [comment]
@Flarna okay, I’ll get to it soon

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lifeiscontent** [comment]
@Flarna @SimonSchick do either of you know how to run tests locally?

**Comment by sandersn** [comment]
@lifeiscontent It should be `npm test`. Let me know what errors you get if that doesn't work.

**Comment by lifeiscontent** [comment]
@sandersn here's my log

```
$ npm test

> definitely-typed@0.0.3 test /Users/aaronreisman/Code/DefinitelyTyped
> node node_modules/types-publisher/bin/tester/test.js --run-from-definitely-typed

Clean data
Clean logs
Clean output
Using local Definitely Typed at /Users/aaronreisman/Code/DefinitelyTyped.
Parsing definitions...
Found 5817 packages.
Parsing in parallel...
Error: Unused file /Users/aaronreisman/Code/DefinitelyTyped/types/ol/assertionerror/index.d.ts
    at checkAllUsedRecur (/Users/aaronreisman/Code/DefinitelyTyped/node_modules/types-publisher/src/lib/definition-parser.ts:422:23)
    at <anonymous>
Error: Parsing failed.
    at fail (/Users/aaronreisman/Code/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:311:20)
    at ChildProcess.child.on (/Users/aaronreisman/Code/DefinitelyTyped/node_modules/types-publisher/src/util/util.ts:299:21)
    at emitNone (events.js:106:13)
    at ChildProcess.emit (events.js:208:7)
    at finish (internal/child_process.js:747:14)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickCallback (internal/process/next_tick.js:181:9)
npm ERR! Test failed.  See above for more details.
```

**Comment by sandersn** [comment]
@lifeiscontent Are you running on Linux or Windows? Here's a commit from February that changed the case of assertionerror to AssertionError. That could be the problem, if the case change failed to do anything on Windows.

I'm not sure what is the command to force sync a part of the repo. You might also try `npm test` on a fresh clone of DefinitelyTyped to see if it passes there.

**Comment by lifeiscontent** [comment]
@sandersn I'm on a Mac, I'll try a fresh clone now.

**Comment by lifeiscontent** [comment]
@sandersn that fixed it, thank you!

**Comment by lifeiscontent** [comment]
@sandersn @Flarna @SimonSchick I'm almost finished here,

it seems like the node-backoff definition is using a generic for the error argument and I guess generics don't merge well with union types.

any suggestions on how to move forward on that?

**Comment by Flarna** [comment]
I think you have to add `| null` also there, e.g.:
```
export interface BackoffInstance {
	<A>(
		fn: (callback: (a: A | null) => void) => void,
		callback: (a: A | null, b: null, priorErrors: A[] | undefined) => void): void;
	<A>(
		fn: (callback: (a: A | null) => void) => void,
		intermediate: BackoffIntermediate<A | null>,
		callback: (a: A | null, b: null, priorErrors: A[] | undefined) => void): void;
	<A, P0>(
		fn: (p0: P0, callback: (a: A | null) => void) => void,
		p0: P0,
		callback: (a: A | null, b: null, priorErrors: A[] | undefined) => void): void;
	<A, P0>(
		fn: (p0: P0, callback: (a: A | null) => void) => void,
		p0: P0,
		intermediate: BackoffIntermediate<A | null>,
		callback: (a: A | null, b: null, priorErrors: A[] | undefined) => void): void;
```

But not sure if this is really correct. Maybe ping also the owners of this definition for a review.

**Comment by lifeiscontent** [comment]
@BendingBender any advice here?

**Comment by lifeiscontent** [comment]
@sandersn @Flarna @SimonSchick in some cases are errors optional? I'm wondering if it's just 1 arg for the callback if I should make `(err?: Error | null)` instead of `(err: Error | null)`

in terms of test, I've got all of them passing except for the ones from nodemailer and hapi, should I just reach out to some of the maintainers of those packages?

**Comment by lifeiscontent** [comment]
@Flarna do you know how to get this merged?

**Comment by Flarna** [comment]
CI is green and it has been approved. There is a dedicated team of maintainers which merge such PRs usually within 1-2 days. Once merged a new NPM package is published automatically.

**Comment by Flarna** [comment]
> I'm wondering if it's just 1 arg for the callback if I should make `(err?: Error | null)` instead of `(err: Error | null)`

I recommend to look into the source code to get some hints. As users usually write something like `if (err) ` it's don't care so I think `(err?: Error | null)` should be ok. But as you have seen such simple changes often result in quite some side effects in dependent modules....

**Comment by jstewmon** [comment]
This broke `promisify` declarations - they no longer support passing `null` for the `err` argument. :-(

**Comment by lifeiscontent** [comment]
@jstewmon lets fix it then! Can you open a PR for the test case?

**Comment by jstewmon** [comment]
There was a test case for `cbOptionalError` in node/test/util.ts which broke, but was erroneously modified to pass.

I opened #35018 to restore things, and update the other promisify tests to include `| null` for consistency.

**Comment by lifeiscontent** [comment]
@jstewmon great! Nice work 👏

**Pull request merged**
